File: arrays/42. Trapping Rain Water/trap.py
class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        amount = 0
        left = 0
        right = len(height) - 1
        left_max_height = height[left]
        right_max_height = height[right]

        while left < right:
            if height[left] < height[right]:
                left_max_height = max(left_max_height, height[left])
                amount += left_max_height - height[left]
                left += 1
            else:
                right_max_height = max(right_max_height, height[right])
                amount += right_max_height - height[right]
                right -= 1
        
        return amount

File: arrays/42. Trapping Rain Water/trap.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let amount = 0;
    let left = 0;
    let right = height.length - 1;
    let leftMaxHeight = height[left];
    let rightMaxHeight = height[right];

    while (left < right) {
        if (height[left] < height[right]) {
            leftMaxHeight = Math.max(leftMaxHeight, height[left]);
            amount += leftMaxHeight - height[left];
            left++;
        } else {
            rightMaxHeight = Math.max(rightMaxHeight, height[right]);
            amount += rightMaxHeight - height[right];
            right--;
        }
    }

    return amount;
};